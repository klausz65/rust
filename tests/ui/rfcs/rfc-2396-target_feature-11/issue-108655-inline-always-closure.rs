// Tests #108655: closures in `#[target_feature]` functions can still be marked #[inline(always)]

//@ check-pass
//@ only-x86_64

#![feature(target_feature_11)]

pub fn okay() {
    ({
        #[inline(always)]
        move || {}
    })();
}

#[target_feature(enable = "avx")]
pub unsafe fn also_okay() {
    ({
        #[inline]
        move || {}
    })();
}

#[target_feature(enable = "avx")]
pub unsafe fn warn() {
    ({
        #[inline(always)]
        //~^ WARNING: cannot use `#[inline(always)]` with `#[target_feature]` [inline_always_closure_in_target_feature_function]
        //~^^ WARNING: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
        move || {}
    })();
}

#[target_feature(enable = "avx")]
pub unsafe fn also_warn() {
    ({
        move || {
            ({
                #[inline(always)]
                //~^ WARNING: cannot use `#[inline(always)]` with `#[target_feature]` [inline_always_closure_in_target_feature_function]
                //~^^ WARNING: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
                move || {}
            })();
        }
    })();
}

fn main() {}
