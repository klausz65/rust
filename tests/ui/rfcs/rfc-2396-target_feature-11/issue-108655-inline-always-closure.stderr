warning: cannot use `#[inline(always)]` with `#[target_feature]`
  --> $DIR/issue-108655-inline-always-closure.rs:26:9
   |
LL |         #[inline(always)]
   |         ^^^^^^^^^^^^^^^^^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #108655 <https://github.com/rust-lang/rust/issues/108655>
   = note: `#[target_feature]` will not be applied to this closure and may result in unpredictable code generation
   = help: consider using `#[inline]` instead
   = note: `#[warn(inline_always_closure_in_target_feature_function)]` on by default

warning: cannot use `#[inline(always)]` with `#[target_feature]`
  --> $DIR/issue-108655-inline-always-closure.rs:38:17
   |
LL |                 #[inline(always)]
   |                 ^^^^^^^^^^^^^^^^^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #108655 <https://github.com/rust-lang/rust/issues/108655>
   = note: `#[target_feature]` will not be applied to this closure and may result in unpredictable code generation
   = help: consider using `#[inline]` instead

warning: 2 warnings emitted

